src/shogun/classifier/Perceptron.cpp

```diff
@@ -65,15 +67,19 @@ bool CPerceptron::train_machine(CFeatures* data)
 	ASSERT(num_vec==train_labels.vlen)
 	SGVector<float64_t> output(num_vec);
 
-	SGVector<float64_t> w = get_w();
+	auto w = SGVector<float64_t>(num_feat);
+	set_w(w);
```

Comment:
this is bad.
You over-write the previous w in any case, which kills the feature of users being able to initialise themselves.
The setting should go into the else branch of the if statement.